File: packages/native-android/wrst-runtime/src/main/java/com/wrst/runtime/SocketClient.kt
package com.wrst.runtime

import android.os.Build
import android.util.Log
import okhttp3.*
import org.json.JSONObject

// The dev server host as seen from a Wear OS emulator (10.0.2.2 aliases the host
// machine's localhost). Physical-device support comes with the CLI (Phase H/I).
const val EMULATOR_HOST_IP = "10.0.2.2"

class SocketClient {

    private val client = OkHttpClient()
    private var webSocket: WebSocket? = null

    private var onMessageReceived: ((String) -> Unit)? = null

    fun setListener(listener: (String) -> Unit) {
        onMessageReceived = listener
    }

    fun connect() {
        val request = Request.Builder()
            .url("ws://${EMULATOR_HOST_IP}:8082")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {

            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d("SocketClient", "[ws] connected")
                // Identify this device to the dev server (for the `l` key binding).
                val hello = JSONObject().apply {
                    put("type", "hello")
                    put("name", "${Build.MANUFACTURER} ${Build.MODEL}")
                }
                webSocket.send(hello.toString())
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                Log.d("SocketClient", "[ws] message: $text")
                // Server only sends a re-pull nudge; the HTTP response is authoritative.
                pullCode()
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.d("SocketClient", "[ws] closing: $code / $reason")
                webSocket.close(1000, null)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d("SocketClient", "[ws] closed: $code / $reason")
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e("SocketClient", "[ws] error: ${t.message}")
                ErrorHandler.set("[ws] error: ${t.message}")
            }
        })
    }

    fun pullCode() {
        val request = Request.Builder()
            .url("http://${EMULATOR_HOST_IP}:8081/bundle.js")
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: java.io.IOException) {
                Log.e("SocketClient", "[http] fetch failed: ${e.message}")
                ErrorHandler.set("[http] fetch failed: ${e.message}")
            }

            override fun onResponse(call: Call, response: Response) {
                Log.d("SocketClient", "onResponse: ${response.code}")
                if (response.code == 503) {
                    // No bundle built yet - wait for a build
                    Log.d("SocketClient", "[http] bundle not ready, waiting for build")
                    return
                }

                val body = response.body.string()
                if (response.code == 422) {
                    // Server is hosting a build error instead of the bundle
                    ErrorHandler.set(if (body.isNotEmpty()) body else "Build failed")
                    return
                }
                if (!response.isSuccessful) {
                    Log.e("SocketClient", "[http] bad response: ${response.code}")
                    ErrorHandler.set("[http] bad response: ${response.code}")
                    return
                }

                Log.d("SocketClient", "[http] bundle fetched (${body.length} chars)")
                ErrorHandler.set(null);
                onMessageReceived?.invoke(body)
            }
        })
    }

    fun send(message: String) {
        webSocket?.send(message)
    }

    fun disconnect() {
        webSocket?.close(1000, "bye")
    }
}

File: packages/wrst/ios/wrst-runtime/Sources/WrstRuntime/SocketClient.swift
import Foundation
import WatchKit

// pulls the bundle over HTTP and listens for reload /
// build_error signals over a WebSocket. Host is "localhost" because the watchOS
// simulator shares the Mac's network (Android used 10.0.2.2 for the emulator).
@MainActor
final class SocketClient {
    private let host: String
    private let session = URLSession(configuration: .default)
    private var webSocket: URLSessionWebSocketTask?

    var onCode: (String) -> Void = { _ in }
    var onError: (String) -> Void = { _ in }

    init(host: String = "localhost") {
        self.host = host
    }

    func connect() {
        guard let url = URL(string: "ws://\(host):8082") else { return }
        let task = session.webSocketTask(with: url)
        webSocket = task
        task.resume()
        sendHello()
        receive()
    }

    // Identify this device to the dev server (for the `l` key binding).
    private func sendHello() {
        let name = WKInterfaceDevice.current().name
        guard let data = try? JSONSerialization.data(withJSONObject: ["type": "hello", "name": name]),
              let json = String(data: data, encoding: .utf8) else { return }
        webSocket?.send(.string(json)) { _ in }
    }

    func disconnect() {
        webSocket?.cancel(with: .goingAway, reason: nil)
        webSocket = nil
    }

    private func receive() {
        webSocket?.receive { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                switch result {
                case .success(let message):
                    if case .string(let text) = message { self.handle(text) }
                    self.receive()
                case .failure(let error):
                    self.onError("[ws] error: \(error.localizedDescription)")
                }
            }
        }
    }

    private func handle(_ text: String) {
        // Server only sends a re-pull nudge; the HTTP response is authoritative.
        pullCode()
    }

    func pullCode() {
        guard let url = URL(string: "http://\(host):8081/bundle.js") else { return }
        let task = session.dataTask(with: url) { [weak self] data, response, error in
            Task { @MainActor in
                guard let self else { return }
                if let error {
                    self.onError("[http] fetch failed: \(error.localizedDescription)")
                    return
                }
                let status = (response as? HTTPURLResponse)?.statusCode ?? -1
                if status == 503 { return } // no bundle built yet - wait for build
                let body = data.flatMap { String(data: $0, encoding: .utf8) }
                if status == 422 {
                    // server is hosting a build error instead of the bundle
                    self.onError(body ?? "Build failed")
                    return
                }
                guard (200..<300).contains(status), let body else {
                    self.onError("[http] bad response: \(status)")
                    return
                }
                self.onCode(body)
            }
        }
        task.resume()
    }
}
